diff --git a/drivers/sensor/nordic/npm1300_charger/npm1300_charger.c b/drivers/sensor/nordic/npm1300_charger/npm1300_charger.c
index c6cfd8cbe2e..3e5cfd620ba 100644
--- a/drivers/sensor/nordic/npm1300_charger/npm1300_charger.c
+++ b/drivers/sensor/nordic/npm1300_charger/npm1300_charger.c
@@ -596,6 +596,11 @@ int npm1300_charger_init(const struct device *dev)
 		}
 	}
 
+	ret = mfd_npm1300_reg_write(config->mfd, CHGR_BASE, CHGR_OFFSET_DIS_SET, 2U);
+	if (ret != 0) {
+		return ret;
+	}
+
 	/* Enable charging if configured */
 	if (config->charging_enable) {
 		ret = mfd_npm1300_reg_write(config->mfd, CHGR_BASE, CHGR_OFFSET_EN_SET, 1U);
