diff --git a/drivers/sensor/npm1300_charger/npm1300_charger.c b/drivers/sensor/npm1300_charger/npm1300_charger.c
index 95634a0a50..7495e0f53e 100644
--- a/drivers/sensor/npm1300_charger/npm1300_charger.c
+++ b/drivers/sensor/npm1300_charger/npm1300_charger.c
@@ -484,6 +484,11 @@ int npm1300_charger_init(const struct device *dev)
 		}
 	}
 
+	ret = mfd_npm1300_reg_write(config->mfd, CHGR_BASE, CHGR_OFFSET_DIS_SET, 2U);
+	if (ret != 0) {
+		return ret;
+	}
+
 	/* Enable charging if configured */
 	if (config->charging_enable) {
 		ret = mfd_npm1300_reg_write(config->mfd, CHGR_BASE, CHGR_OFFSET_EN_SET, 1U);