diff --git a/drivers/display/display_sdl.c b/drivers/display/display_sdl.c
index 2a457249ed..ed624bb2c0 100644
--- a/drivers/display/display_sdl.c
+++ b/drivers/display/display_sdl.c
@@ -54,7 +54,7 @@ static int sdl_display_init(const struct device *dev)
 #endif /* SDL_DISPLAY_DEFAULT_PIXEL_FORMAT */
 		;
 
-	int rc = sdl_display_init_bottom(config->height, config->width, &disp_data->window,
+	int rc = sdl_display_init_bottom(config->height * 4, config->width * 4, &disp_data->window,
 					 &disp_data->renderer, &disp_data->texture);
 
 	if (rc != 0) {
